src/containers/Comments/Download/hooks.js

```diff
@@ -15,3 +25,120 @@ export function useDownloadComments(token) {
 
   return { comments, onFetchCommentsTimestamps };
 }
+
+const TIMESTAMPS_PAGE_SIZE = 100;
+export function useDownloadCommentsTimestamps(recordToken) {
+  const { apiInfo } = useLoader();
+  const [timestamps, setTimestamps] = useState(null);
+  const [remaining, setRemaining] = useState([]);
+  const [progress, setProgress] = useState(0);
+  const commentsSelector = useMemo(
+    () => sel.makeGetRecordComments(recordToken),
+    [recordToken]
+  );
+  const comments = useSelector(commentsSelector);
+  const commentsLength = comments.length || 0;
+  const onFetchCommentsTimestamps = useAction(act.onFetchCommentsTimestamps);
+
+  useEffect(() => {
+    const ids = comments && comments.map((c) => c.commentid);
+    setRemaining(ids);
+  }, [comments]);
+
+  const getCommentIdsForPagination = (commentIds) => [
+    take(TIMESTAMPS_PAGE_SIZE)(commentIds),
+    takeRight(commentIds.length - TIMESTAMPS_PAGE_SIZE)(commentIds)
+  ];
+
+  const [
+    state,
+    send,
+    { START, VERIFY, FETCH, RESOLVE, REJECT }
+  ] = useFetchMachine({
+    actions: {
+      initial: () => {
+        const ts = loadCommentsTimestamps(recordToken);
+        const loaded = ts && Object.keys(ts.comments).length === commentsLength;
+        if (loaded) {
+          return send(RESOLVE, { timestamps: ts });
+        }
+        if (recordToken && !timestamps) {
+          return send(START);
+        }
+        return;
+      },
+      start: () => {
+        // fetch first page of comments timestamps
+        const [fetch, next] = getCommentIdsForPagination(remaining);
+        onFetchCommentsTimestamps(recordToken, fetch)
+          .then(({ comments }) => {
+            setTimestamps({ comments });
+            setProgress(
+              ((Object.keys(comments).length * 100) / commentsLength).toFixed(2)
+            );
+            setRemaining(next);
+            return send(VERIFY);
+          })
+          .catch((e) => send(REJECT, e));
+        return send(FETCH);
+      },
+      verify: () => {
+        if (remaining.length === 0) {
+          // all timestamps loaded, resolve
+          handleSaveCommentsTimetamps(recordToken, timestamps);
+          setProgress(100);
+          fileDownload(
+            JSON.stringify(
+              {
+                comments: timestamps.comments,
+                serverpublickey: apiInfo.pubkey
+              },
+              null,
+              2
+            ),
```

Comment:
I think you could move it to another function too, what do you think? Little hard to understand what's going on here just by looking at it :)